---
title: "Whole transcriptomic network analysis using Co-expression Differential Network Analysis (CoDiNA)"
date: 2020-10-15
date_end: '2020-10-15'
event: 'Plos One'
event_url: 'https://doi.org/10.1371/journal.pone.0240523'
excerpt: "Here we present a method for the systematic comparison of an unlimited number of networks, with unlimited number of transcripts: Co-expression Differential Network Analysis (CoDiNA)."
show_post_time: false
links:
- icon: newspaper
  icon_pack: fas
  name: publication
  url: https://doi.org/10.1371/journal.pone.0240523
author: "Deisy Morselli Gysi"
draft: false
# layout options: single, single-sidebar
layout: single
categories:
- paper
- co-expression
- r package
- networks
---

# Abstract
Biological and medical sciences are increasingly acknowledging the significance of gene co-expression-networks for investigating complex-systems, phenotypes or diseases. Typically, complex phenotypes are investigated under varying conditions. While approaches for comparing nodes and links in two networks exist, almost no methods for the comparison of multiple networks are available and—to best of our knowledge—no comparative method allows for whole transcriptomic network analysis. However, it is the aim of many studies to compare networks of different conditions, for example, tissues, diseases, treatments, time points, or species. Here we present a method for the systematic comparison of an unlimited number of networks, with unlimited number of transcripts: Co-expression Differential Network Analysis (CoDiNA). In particular, CoDiNA detects links and nodes that are common, specific or different among the networks. We developed a statistical framework to normalize between these different categories of common or changed network links and nodes, resulting in a comprehensive network analysis method, more sophisticated than simply comparing the presence or absence of network nodes. Applying CoDiNA to a neurogenesis study we identified candidate genes involved in neuronal differentiation. We experimentally validated one candidate, demonstrating that its overexpression resulted in a significant disturbance in the underlying gene regulatory network of neurogenesis. Using clinical studies, we compared whole transcriptome co-expression networks from individuals with or without HIV and active tuberculosis (TB) and detected signature genes specific to HIV. Furthermore, analyzing multiple cancer transcription factor (TF) networks, we identified common and distinct features for particular cancer types. These CoDiNA applications demonstrate the successful detection of genes associated with specific phenotypes. Moreover, CoDiNA can also be used for comparing other types of undirected networks, for example, metabolic, protein-protein interaction, ecological and psychometric networks. CoDiNA is publicly available as an R package in CRAN (https://CRAN.R-project.org/package=CoDiNA).

# Citation

Gysi, D. M., Fragoso, T.M, Zebardast, F., Beroli, W., Busskamp, V.,  Almaas, E., Nowick, K. (2018). Whole transcriptomic network analysis using Co-expression Differential Network Analysis (CoDiNA). Plos One, https://doi.org/10.1371/journal.pone.0240523
